Add Style/LineBreak Cops

panthomakos opened the pull request:
When large multi-line lists are present in a code base it's often
rather arbitrary to align them based on the first line/element:

```
long_variable = { foo: 'foo',
                  bar: 'bar',
                  baz: 'baz' }
```

In this example if the variable name is refactored the entire list
formatting also needs to change. This makes diffs, in particular, very
hard to read due to the large number of formatting-only changes:

```
short = { foo: 'foo',
          bar: 'bar',
          baz: 'baz' }
```

These alignments also greatly reduce the available space on a line for
each element, especially if there is a maximum line length enforcement.

The `Style/MultilineList` cop along with any appropriate Align or Indent
cop (for example `HashAlign` and `HashIndent`) allows you to enforce a
more diff-friendly and short-line style:

```
variable = {
  foo: 'foo',
  bar: 'bar',
  baz: 'baz' }
```

Mutually-exclusive configuration options are provided for each of method
parameters, method arguments, array elements and hash elements. Care has
been taken to ensure that DSL-like method calls are ignored and do not
result in invalid Ruby.

I have added tests for all use-cases I could find in some large Ruby and
Rails code-bases.

jonas054 commented on lib/rubocop/cop/style/multiline_list.rb:
```diff
@@ -0,0 +1,184 @@
+      class MultilineList < Cop
+        MSG = 'Move all elements below the first line of a multi-line list.'
+
+        def on_def(node)
```
You need to handle `defs` nodes too, and you can do this by using the `OnMethodDef` mixin.

jonas054 commented on lib/rubocop/cop/style/multiline_list.rb:
```diff
@@ -0,0 +1,184 @@
+          # Check if we are using an implicit hash.
+          source = node.loc.expression.source_line[node.loc.column..-1]
+
+          if source =~ /^\s*\{/
```
We usually avoid regexp matching when there's a alternative such as `node.loc.begin` for detecting braces in this case.

jonas054 commented on lib/rubocop/cop/style/multiline_list.rb:
```diff
@@ -0,0 +1,184 @@
+      #       :b]
+      #
+      class MultilineList < Cop
+        MSG = 'Move all elements below the first line of a multi-line list.'
```
In the descriptions, and especially in this message, I think it would be more clear to talk about a preceding line break. Something like "Add a line break before the first element...".

jonas054 commented on lib/rubocop/cop/style/multiline_list.rb:
```diff
@@ -0,0 +1,184 @@
+          ['[', '%w', '%W'].any? { |a| source.start_with?(a) }
+        end
+
+        def assignment?(node, limit)
```
I don't think the name reflects what the method does.

jonas054 commented on lib/rubocop/cop/style/multiline_list.rb:
```diff
@@ -0,0 +1,184 @@
+          source =~ /\s*\=\s*$/
+        end
+
+        def parens?(node, limit)
```
Here too the name only reveals part of the story.

panthomakos commented:
jonas054 Thank you very much for the review. I've addressed all of your comments and amended my commit. The autocorrect method is also a lot simpler because I am using the first element as the offense (instead of the parent node) which means inserting a newline is much simpler. PTAL

jonas054 commented on config/default.yml:
```diff
@@ -525,6 +525,13 @@ Style/MethodName:
     - snake_case
     - camelCase
 
+Style/MultilineList:
```
The cop is enabled by default but each part is disabled by default. I think it would be better to do it the other way around, because then people can find it when they look in `disabled.yml` for extra things to turn on.

jonas054 commented on lib/rubocop/cop/style/multiline_list.rb:
```diff
@@ -0,0 +1,167 @@
+          check_children(node, children)
+        end
+
+        def explicit_array?(source)
```
With arrays too, out can just check if `node.loc.begin` is "truthy".

jonas054 commented on lib/rubocop/cop/style/multiline_list.rb:
```diff
@@ -0,0 +1,167 @@
+          return unless cop_config['EnableHashElements']
+
+          # Check if we are using an implicit hash.
+          if node.loc.begin && node.loc.begin.is?('{')
```
The `&& node.loc.begin.is?('{')` is not necessary. It might add some clarity, but the specs still pass if you remove it.

panthomakos replied:
I am now understanding now `node.loc.begin` works. Thanks for the comments.

jonas054 commented on lib/rubocop/cop/style/multiline_list.rb:
```diff
@@ -0,0 +1,167 @@
+          if node.loc.begin && node.loc.begin.is?('{')
+            check_children(node, node.children)
+          else
+            return unless method_uses_parens?(node.parent, node)
```
I would go for `elsif` rather than `else return unless`.

jonas054 commented on lib/rubocop/cop/style/multiline_list.rb:
```diff
@@ -0,0 +1,167 @@
+        end
+
+        def assignment_on_same_line?(node)
+          source = node.loc.expression.source_line[0...node.loc.column]
```
To avoid regexp matching we can write:
```ruby
_lhs, rhs = *node.parent
node.parent.loc.operator.line == rhs.loc.line
```

jonas054 commented on lib/rubocop/cop/style/multiline_list.rb:
```diff
@@ -0,0 +1,167 @@
+          source =~ /\s*\=\s*$/
+        end
+
+        def method_uses_parens?(node, limit)
```
This one is hard to rewrite so that it doesn't use regexp matching, because it handles both method definitions (two kinds) and method calls. So either leave it as it is, or break it up into two separate methods.

jonas054 commented on lib/rubocop/cop/style/multiline_list.rb:
```diff
@@ -0,0 +1,167 @@
+          source =~ /\s*\(\s*$/
+        end
+
+        def check_children(node, children, start = nil)
```
If you change the default value to `start = node` you can just use `start` instead of `(start || node)` further down.

jonas054 commented on spec/rubocop/cop/style/multiline_list_spec.rb:
```diff
@@ -0,0 +1,419 @@
+        inspect_source(cop, source)
+
+        expect(cop.offenses.length).to eq(1)
+        expect(cop.offenses.first.line).to eq(1)
```
You can add `expect(cop.highlights).to eq(['bar'])` at least in some places where offenses are detected, so we see that the right piece of code is highlighted in the reporting.

jonas054 commented:
That's much better. I've added some more comments, mostly because I didn't get around to do it earlier.

panthomakos commented:
@jonas054 I've addressed your comments. Thanks again for the review. I'm learning a lot about Ruby code parsing :)

PTAL

panthomakos commented:
@jonas054 I am taking a look at some issues w/ the array assignment part. After making your changes I am getting some infinite loops for some implicit arrays in setters. I am investigating to see if I can write a simple test case and resolve the issue.

panthomakos commented on lib/rubocop/cop/style/multiline_list.rb:
```diff
@@ -0,0 +1,158 @@
+
+        def assignment_on_same_line?(node)
+          source = node.loc.expression.source_line[0...node.loc.column]
+          source =~ /\s*\=\s*$/
```
@jonas054 I was unable to convert this using the `lhs, rhs` code you suggested because it does not work for some "assignment" types. It does work for `masgn` but not for a `send` type. For example:

```
config.values =
  :a,
  :b,
  :c
```

jonas054 replied:
I see. Well, both `assignment_on_same_line?` and `method_uses_parens?` handle more than one node type, and then you have to do it with regexp matching or add handling of the different cases. I guess it's OK as it is.

panthomakos commented:
@jonas054 I resolved the issue w/ implicit arrays and added a test case to demonstrate the issue. PTAL

panthomakos commented on lib/rubocop/cop/style/multiline_list.rb:
```diff
@@ -0,0 +1,158 @@
+          return if children.size < 2
+
+          line = start.loc.line
+          min = children.min_by { |n| n.loc.line }
```
@jonas054 I have a question related to this code. Is there a standard way you check for "does this node span multiple lines"? For example, the following test case won't pass this implementation:

```
something('foo', bar: 1,
  baz: 2)
```

The reason is that even though there are two children, they both start on the same line. While I could make some kind of check to see if any of the individual children have a "\n" in their source, it seems like there might be a better way to address this issue. Do you know if there is some way I can go about check this so that even these cases are tagged as offenses?

rrosenblum replied:
Assuming that the node has a `begin` and an `end`, which I think it will in this case, you can use `node.loc.begin.line` and `node.loc.end.line`. The same will hold true for other attributes off of `loc`.

rrosenblum commented on lib/rubocop/cop/style/multiline_list.rb:
```diff
@@ -0,0 +1,158 @@
+
+          if node.loc.begin
+            check_children(node, node.children)
+          elsif !method_uses_parens?(node.parent, node)
```
Can the `elsif` and the `else` be flipped here? Assuming this doesn't break anything, I think it would be more readable as

```ruby
elsif method_uses_parens?(node.parent, node)
  check_children(node, node.children, node.parent)
else
  return
end
```

panthomakos commented:
@rrosenblum I've incorporated your comments. I resolved the issue with elements spanning multiple lines by using the `first_line` and `last_line` of the `loc` object instead of just the `line` number. I've added a test case that passes with the updated code as well.

jonas054 commented on lib/rubocop/cop/style/multiline_list.rb:
```diff
@@ -0,0 +1,158 @@
+            check_children(node, node.children)
+          elsif method_uses_parens?(node.parent, node)
+            check_children(node, node.children, node.parent)
+          else
```
`else return` isn't really needed.

panthomakos commented:
Incorporated comments. Removed the `else return`.

jonas054 commented:
:+1: Looks good!

panthomakos commented:
Anything else I need to do to get this merged?

panthomakos commented:
Rebased off master to resolve a conflict in `CHANGELOG.md`. Is there anything else that needs to be done in this PR before it is merged?

panthomakos commented:
ping @jonas054 @rrosenblum I rebased off master again. Is there anything else I need to do here to get this change merged to master?

rrosenblum commented:
@panthomakos the build is failing for an incorrect format in the CHANGELOG. There is a missing colon after the link to the PR.

Should be:
```
* [#2277](https://github.com/bbatsov/rubocop/pull/2277):
```

panthomakos commented:
@rrosenblum fixed the CHANGELOG formatting

rrosenblum commented on spec/rubocop/cop/style/multiline_list_spec.rb:
```diff
@@ -0,0 +1,483 @@
+      it 'autocorrects the offense' do
+        new_source = autocorrect_source(cop, source)
+
+        expect(new_source).to eq(
```
This doesn't seem like the correct auto-correction to me. The example given in the documentation comment shows
```ruby
# good with EnableMethodParams
def method(
  foo, bar,
  baz)
  do_something
end
```

The auto-correct places the first element left justified with the `def`. 
```ruby
def foo(
bar,
  baz)
  do_something
end
```

Should it be indented or is it relying on another cop to handle that indention?

panthomakos replied:
It relies on `Style/FirstParameterIndetation`:

http://www.rubydoc.info/gems/rubocop/0.29.1/RuboCop/Cop/Style/FirstParameterIndentation

panthomakos replied:
Any additional alignment should then be fixed by `Style/AlignParameters`:

http://www.rubydoc.info/gems/rubocop/0.29.1/RuboCop/Cop/Style/AlignParameters

bbatsov commented:
I haven't had time to review the code, but I can tell you that I definitely don't like the name of the cop, as it's not clear what it refers to. Certainly, the term "list" is not used often in Ruby.

panthomakos commented:
@bbatsov I suppose the other option here is to have 4 separate cops that indicate that a line break is necessary, something like: `FirstHashElementLinePlacement` or `FirstArrayElementLineBreak`. Maybe I can generalize it to `FirstMultilineElementLinePlacement`?

bbatsov commented:
Probably this should be a few separate cops. @jonas054 Thoughts?

jonas054 commented:
I hadn't thought about that while reviewing, but sure, why not? It's more simple in a way to have four separate cops than one cop with four enable properties. And the naming should be more natural.

panthomakos commented:
@bbatsov @jonas054 

Please take another look. I've split the cop up into 4 separate cops with cleaner descriptions and messages.

jonas054 commented:
:+1: Looks good to me.

bbatsov commented:
Looks good. Just remove the `.` in the commit message's title.

Btw, there should probably a set of complementary checks for the final parenthesis, so you can enforce either this style:

```
variable = {
  foo: 'foo',
  bar: 'bar',
  baz: 'baz' }
```

or this one (which is my preference):

```
variable = {
  foo: 'foo',
  bar: 'bar',
  baz: 'baz' 
}
```

panthomakos commented:
@bbatsov Removed period from commit message.

I prefer the second as well.  I was thinking about adding that set of cops after this PR was complete.

bbatsov commented:
> I was thinking about adding that set of cops after this PR was complete.

Looking forward to those. :-)
